lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java

```diff
@@ -45,54 +45,89 @@
       final String containerId = generateContainerId();
       this.containerRequests.put(containerId, resourceRequestEvent);
+      this.containerCount.incrementAndGet();
+      this.azureBatchTaskStatusAlarmHandler.enableAlarm();
       this.evaluatorShimManager.onResourceRequested(containerId, resourceRequestEvent);
     }
   }
 
   public void onResourceReleased(final ResourceReleaseEvent resourceReleaseEvent) {
-    LOG.log(Level.FINEST, "Got ResourceReleasedEvent for Id: {0} in AzureBatchResourceManager",
-        resourceReleaseEvent.getIdentifier());
-    this.containerRequests.remove(resourceReleaseEvent.getIdentifier());
+    String id = resourceReleaseEvent.getIdentifier();
+    LOG.log(Level.FINEST, "Got ResourceReleasedEvent for Id: {0} in AzureBatchResourceManager", id);
+    if (!this.activeContainerIds.remove(id)) {
+      LOG.log(Level.WARNING,
+          "Attempting to remove non-existent activeContainer for Id: {0} in AzureBatchResourceManager", id);
+    }
+
+    ResourceRequestEvent removedEvent = this.containerRequests.remove(id);
+    if (removedEvent == null) {
+      LOG.log(Level.WARNING,
+          "Ignoring attempt to remove non-existent containerRequest for Id: {0} in AzureBatchResourceManager", id);
+    }
+    int currentContainerCount = this.containerCount.decrementAndGet();
+    if (currentContainerCount <= 0) {
+      this.azureBatchTaskStatusAlarmHandler.disableAlarm();
+    }
     this.evaluatorShimManager.onResourceReleased(resourceReleaseEvent);
   }
 
   public void onResourceLaunched(final ResourceLaunchEvent resourceLaunchEvent) {
-    final int evaluatorMemory = this.containerRequests.get(resourceLaunchEvent.getIdentifier()).getMemorySize().get();
+    String id = resourceLaunchEvent.getIdentifier();
+    LOG.log(Level.FINEST, "Got ResourceLaunchEvent for Id: {0} in AzureBatchResourceManager", id);
+    final int evaluatorMemory = this.containerRequests.get(id).getMemorySize().get();
     String launchCommand = this.launchCommandBuilder.buildEvaluatorCommand(resourceLaunchEvent,
         evaluatorMemory, this.jvmHeapFactor);
     String evaluatorConfigurationString = this.configurationSerializer.toString(resourceLaunchEvent.getEvaluatorConf());
-
+    this.activeContainerIds.add(id);
     // TODO[At this point we are only sending the evaluator config and launch command to the shim. AllocatedEvaluator]
     // TODO[object also has "addFile" and "addLibrary" methods, so maybe we'll need to package the JAR for the]
     // TODO[evaluator, too, and not assume that both Evaluator and Shim can share the same JAR.]
     this.evaluatorShimManager.onResourceLaunched(resourceLaunchEvent, launchCommand, evaluatorConfigurationString);
   }
 
+  public Boolean isContainerActive(final String containerId) {
+    return this.activeContainerIds.contains(containerId);
+  }
+
+  public String activeContainerList() {
+    return String.join(", ", this.activeContainerIds);
```

Comment:
I'm not sure about the thread safety aspect of this. Java documentation states that synchronization must be used when iterating over the elements (https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedSet(java.util.Set)). One way to address this would be to copy the set into a stack variable. But if this is not critical, I'd get rid of this alltogether. 
Also, according to Java naming convention, this method should be called "getActiveContainerListString".

Reply:
Removed it. I had added it for debugging purposes to test my code.